docs/v1/P4Runtime-Spec.mdk

```diff
@@ -3795,34 +3803,56 @@ the `GetForwardingPipelineConfig RPC`. The request is defined as:
 
 ~ Begin Proto
 message GetForwardingPipelineConfigRequest {
+  enum ResponseType {
+    ALL = 0;
+    COOKIE_ONLY = 1;
+    P4INFO_ONLY = 2;
+  }
   uint64 device_id = 1;
+  ResponseType response_type = 2;
 }
 ~ End Proto
 
 The `device_id` uniquely identifies the target P4 device. An `INVALID_ARGUMENT`
 error is returned if the `device_id` is not recognized by the P4Runtime server.
 
-The response contains the `P4ForwardingPipelineConfig` for the specified device:
+The `response_type` is used to specify which fields to populate in the response,
+its value can be one of:
+
+* `ALL`: returns a `ForwardingPipelineConfig` with all fields
+  set as stored by the target. This is the default behaviour if the
+  `response_type` field is not set.
+
+* `COOKIE_ONLY`: reply by setting only the `cookie` field in the
+  `ForwardingPipelineConfig`, omitting all other fields. This mechanisms can be
+  used by a controller to verify that a config is the expected one, while
+  minimizing the amount of data in the response message.
+
+* `P4INFO_ONLY`: reply by setting only the `p4info` field.
+
+The response contains the `ForwardingPipelineConfig` for the specified device:
 
 ~ Begin Proto
- message GetForwardingPipelineConfigResponse {
+message GetForwardingPipelineConfigResponse {
   ForwardingPipelineConfig config = 1;
 }
 ~ End Proto
 
-The `ForwardingPipelineConfig` consists of P4Info and the opaque
-target-dependent `p4_device_config`. If a P4Runtime server is in a state where
-the forwarding-pipeline config is not known, the top-level `config` field will
-be unset in the response. Examples are (i) a server that only allows
-configuration via `SetFowardingPipelineConfig` but this RPC hasn't been invoked
-yet, (ii) a server that is configured using a different mechanism but this
-configuration hasn't yet occured.
+If a P4Runtime server is in a state where the forwarding-pipeline config is not
+known, the top-level `config` field will be unset in the response. Examples are
+(i) a server that only allows configuration via `SetFowardingPipelineConfig` but
+this RPC hasn't been invoked yet, (ii) a server that is configured using a
+different mechanism but this configuration hasn't yet occured.
 
 Once a forwarding-pipeline config is installed on the device (either via
 `SetFowardingPipelineConfig` or a different mechanism), some P4Runtime servers
 may not support retrieval of the target-dependent config, in which case
-`config.p4_device_config` will be empty / unset in the response. However, all
-P4Runtime servers are required to return P4Info in this scenario.
+`config.p4_device_config` will be empty / unset in the response, even if
+`response_type` in the request was set to `ALL`. However, all P4Runtime servers
+are required to return the P4Info and cookie in this scenario. If the config is
+installed on the device with a mechanism other than
+`SetFowardingPipelineConfig`, the value of the returned cookie is left
```

Comment:
what about: the value of the returned cookie will be the default Protobuf value (0).

Reply:
Cookie in sub-message to clearly identify cases where a cookie is not present.

Reply:
Cookie is now unset if one was not present in the `SetForwardingPipelineConfig` RPC, or if the pipeline has been installed via other mechanisms.